sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/CosmosItemTest.java

```diff
@@ -151,6 +156,103 @@ public void readItem() throws Exception {
+                    ObjectNode.class),
+            Duration.ofNanos(10),
+            fallBackProperties)
+            .map(node -> node.get("id").asText())
+            .block();
+
+        assertThat(timedOutResponse).isEqualTo("justFallback");
+
+        // Just ensure the logging of the soft timeout can finish
+        Thread.sleep(1000);
+    }
+
+    static <T> Mono<T> wrapWithSoftTimeoutAndFallback(
+        Mono<CosmosItemResponse<T>> source,
+        Duration softTimeout,
+        T fallback) {
+
+        // Execute the readItem with transformation to return the json payload
+        // asynchronously with a "soft timeout" - meaning when the "soft timeout"
+        // elapses a default/fallback response is returned but the original async call is not
+        // cancelled, but allowed to complete. This makes it possible to still emit diagnostics
+        // or process the eventually successful call
+        AtomicBoolean timeoutElapsed = new AtomicBoolean(false);
+        return Mono
+            .<T>create(sink -> {
+                source
+                    .subscribeOn(Schedulers.boundedElastic())
+                    .subscribe(
+                        response -> {
+                            if (timeoutElapsed.get()) {
+                                logger.warn(
+                                    "COMPLETED SUCCESSFULLY after timeout elapsed. Diagnostics: {}",
+                                    response.getDiagnostics().toString());
+                            } else {
+                                logger.info("COMPLETED SUCCESSFULLY");
+                            }
+
+                            sink.success(response.getItem());
+                        },
+                        error -> {
+                            final Throwable unwrappedException = Exceptions.unwrap(error);
+                            if (unwrappedException instanceof CosmosException) {
+                                final CosmosException cosmosException = (CosmosException) unwrappedException;
+
+                                logger.error(
+                                    "COMPLETED WITH COSMOS FAILURE. Diagnostics: {}",
+                                    cosmosException.getDiagnostics() != null ?
+                                        cosmosException.getDiagnostics().toString() : "n/a",
+                                    cosmosException);
+                            } else {
+                                logger.error("COMPLETED WITH GENERIC FAILURE", error);
+                            }
+
+                            if (timeoutElapsed.get()) {
+                                // fallback returned already - don't emit unobserved error
+                                sink.success();
+                            } else {
+                                sink.error(error);
+                            }
+                        }
+                    );
+            })
+            .timeout(softTimeout)
```

Comment:
Another concern we have for timeoutAPI is that it will mess up the internal retry/refresh logic, this will also prevents that 👍 

The only concern here is that since internally the request is still executed as it is, so it will still take all the resources to process them, and if customer having a aggressive retry strategy, it may eventually cause high CPU, high latency etc

Reply:
Yes - but if they consume all CPU with the 5 seconds request timeout they use the wrong product.